### PR TITLE
Updated JaCoCo to 0.8.3

### PR DESCRIPTION
**Pull request by DreierF**
- [x] [No new Teamscale findings](https://build.cqse.eu/teamscale/delta.html#input/jacoco-client/?showMergeFindings=true)
- [ ] [Changes are tested adequately](https://build.cqse.eu/teamscale/tests.html#/jacoco-client/)
- [x] USAGE.md updated in case of user-visible changes
- [ ] CHANGELOG.md updated
- [x] build.gradle version number increased according to semantic versioning

### CODE REVIEW

**Comment by DreierF** [comment]
Fixes #44

File: report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingInstructionsBuilder.java [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+	public void fillCache() {
+		// Wire jumps:
+		for (final Jump j : jumps) {
+			j.wire();
+		}
+
+		// ADDED
+		// Traces back all instructions that are executed before reaching a probe
+		// and stores the mapping from probe to lines in #classCoverageLookup
+		for (final CoveredProbe p : coveredProbes) {
```
Comment (karottenreibe) [inline report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingInstructionsBuilder.java]: If this is our code, please don't use unnecessary final and please don't use one-letter variable names

File: report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingInstructionsBuilder.java [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+	 * ADDED
+	 * Helper to get the private field predecessor from an instruction.
+	 */
+	private Instruction getPredecessor(Instruction instruction) {
+		try {
+			Field f = instruction.getClass().getDeclaredField("predecessor");
+			f.setAccessible(true);
+			instruction = (Instruction) f.get(instruction);
+		} catch (NoSuchFieldException | IllegalAccessException e) {
+			e.printStackTrace();
```
Comment (karottenreibe) [inline report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingInstructionsBuilder.java]: handle with proper logging? Also at least please log that this means we have a programming error that we have to handle here in the code (I think users can't do anything in this case, right?)

File: report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingInstructionsBuilder.java [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+		}
+	}
+
+	/**
+	 * ADDED
+	 * Helper to get the private field predecessor from an instruction.
+	 */
+	private Instruction getPredecessor(Instruction instruction) {
+		try {
+			Field f = instruction.getClass().getDeclaredField("predecessor");
```
Comment (karottenreibe) [inline report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingInstructionsBuilder.java]: please don't use one-letter variable names

File: report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingInstructionsBuilder.java [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+				coveredLines.add(instruction.getLine());
+				instruction = getPredecessor(instruction);
+			}
+			classCoverageLookup.addProbe(p.probeId, coveredLines);
+		}
+	}
+
+	/**
+	 * ADDED
+	 * Helper to get the private field predecessor from an instruction.
```
Comment (karottenreibe) [inline report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingInstructionsBuilder.java]: please explain what is stored in the predecessor field

File: report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingInstructionsBuilder.java [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+	 * instances
+	 */
+	public void fillCache() {
+		// Wire jumps:
+		for (final Jump j : jumps) {
+			j.wire();
+		}
+
+		// ADDED
+		// Traces back all instructions that are executed before reaching a probe
```
Comment (karottenreibe) [inline report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingInstructionsBuilder.java]: please explain why we need to trace back predecessors. is this because jacoco only inserts a probe at the end of a linear sequence of instructions?

File: report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingInstructionsBuilder.java [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+	/**
+	 * Adds a new probe for the last instruction.
+	 *
+	 * @param probeId index in the probe array
+	 * @param branch  unique branch number for the last instruction
+	 */
+	void addProbe(final int probeId, final int branch) {
+//		REMOVED check of probes array and instead add the probes unconditionally
+//		final boolean executed = probes != null && probes[probeId];
+		currentInsn.addBranch(true, branch);
```
Comment (karottenreibe) [inline report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingInstructionsBuilder.java]: is this line also added by us? if so please explain what this does and why it's needed. if not, please change the order of this and the next instruction and add a blank line to make it clear that this is unrelated to the comment

**Comment by karottenreibe** [comment]
please update CHANGELOG
how did you test this? I think we should at least do some rudimentary manual test to make sure the caching still works as expected and the resulting coverage is not missing anything or has too much coverage
Or do we have automated tests for this?

File: report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingInstructionsBuilder.java (hunk 2 above)
Comment (DreierF) [reply report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingInstructionsBuilder.java]: I converted into a RuntimeError as we cannot recover from this point anyway.

File: report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingInstructionsBuilder.java (hunk 5 above)
Comment (DreierF) [reply report-generator/src/main/java/org/jacoco/core/internal/analysis/CachingInstructionsBuilder.java]: Correct, added a short additional comment line.

**Comment by DreierF** [comment]
We do have an automated smoke test in JaCoCoTestwiseReportGeneratorTest that still produces the same coverage as before.
However there is a known bug (TS-18106) that the coverage is not correct in all cases, but that will be fixed in a separate ticket. It only affects the line coverage numbers, but not TIA or TGA as a whole as we only consider method level coverage.

**Comment by DreierF** [comment]
RFR
Is there anyway I can signal this via the GitHub UI?

**Comment by karottenreibe** [comment]
Unfortunately, I don't know of any way. We could have a look at https://reviewable.io/ which maybe has a better review workflow than Github. But I haven't had the time to deep-dive into it to check it out. Just from the website I can't tell if this would work for us. If you'd like you can have a look and let me know what you think :-)

**Comment by karottenreibe** [comment]
good work!
